contracts/root/StakeManager.sol

```diff
@@ -3,192 +3,282 @@ pragma solidity ^0.4.24;
+    stakers[msg.sender].deActivationEpoch = currentEpoch + dynasty*2;
+
+    // new descendant selection 
+    uint256 value = nextValidatorList.delMax(); 
+    if (value != 0) {
+      validatorList.insert(value);
+      value >> 160;
+      address newStaker = address(uint160(value));
+      stakers[newStaker].activationEpoch = stakers[msg.sender].deActivationEpoch;
+      emit StakeInit(newStaker, stakers[newStaker].amount, "0x0");
+    }
+    emit UnstakeInit(msg.sender, amount, "0x0");
+  }
+  
+  function getCurrentValidatorSet() public view returns (address[]) {
+    return validatorList.getTree();
+  }
 
-    // transfer stake amount
-    tokenObj.transfer(msg.sender, amount);
-
-    // broadcast event
-    emit Unstaked(msg.sender, amount, stakedAmount, data);
+  function getNextValidatorSet() public view returns (address[]) {
+    return nextValidatorList.getTree();
   }
 
-  function totalStakedFor(address addr) public view returns (uint256) {
-    return stakeList[stakers[addr]].amount;
+  function totalStakedFor(address addr) public view returns (uint256) { // onlyowner ?
+    require(addr != address(0x0));
+    return stakers[addr].amount;
   }
 
   function totalStaked() public view returns (uint256) {
     return totalStake;
   }
 
+  function updateEpoch() public { // onlyrootchain /temp function
+    currentEpoch = currentEpoch.add(1);
+  }
+
   function token() public view returns (address) {
     return address(tokenObj);
   }
 
   function supportsHistory() public pure returns (bool) {
     return false;
   }
-
-  // optional
-  // function lastStakedFor(address addr) public view returns (uint256);
-  // function totalStakedForAt(address addr, uint256 blockNumber) public view returns (uint256);
-  // function totalStakedAt(uint256 blockNumber) public view returns (uint256);
-
-  //
-  // PoS functions
-  //
+  
+  function validatorThreshold() public view returns (uint256) {
+    return _validatorThreshold;
+  }
 
   // Change the number of validators required to allow a passed header root
-  function updateValidatorThreshold(uint256 newThreshold) public onlyOwner {
+  function updateValidatorThreshold(uint256 newThreshold) public{ // onlyRootChain {
     emit ThresholdChange(newThreshold, _validatorThreshold);
     _validatorThreshold = newThreshold;
   }
 
-  // Sample a proposer. Likelihood of being chosen is proportional to stake size.
-  function getProposer() public view returns (address) {
-    // Convert the seed to an index
-    uint256 target = uint256(epochSeed) % totalStake;
-    // Index of stake list
-    uint64 i = 1;
-    // Total stake
-    uint256 sum = 0;
-    while (sum < target) {
-      sum += stakeList[i].amount;
-      i += 1;
-    }
+  function updateDynastyValue(uint256 newDynasty) public{ // onlyRootChain {
+    emit DynastyValueChange(newDynasty, dynasty);
+    dynasty = newDynasty;
+  }
+  
+  function getProposer() public  returns (address) {
```

Comment:
remove if not required

Reply:
yeah now will remove whole proposer selection mechanism itself :D.